### PR TITLE
Updated call_api in Upload and Admin to send either callback or promise

### PR DESCRIPTION
**Pull request by strausr**

### CODE REVIEW

File: lib-es5/api_client/execute_request.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ var extend = utils.extend,
 function execute_request(method, params, auth, api_url, callback) {
   var options = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : {};
 
+  var didUserPassCB = typeof callback === "function";
```
[the remark this answers is not in the document]
Comment (asisayag2) [reply lib-es5/api_client/execute_request.js]: Following our discussion yesterday, let's drop this variable completely

File: lib-es5/api_client/execute_request.js [hunk 2]
```diff
@@ -101,9 +103,10 @@ function execute_request(method, params, auth, api_url, callback) {
             query_params
           }
         };
-        deferred.reject(err_obj.error);
-        if (typeof callback === "function") {
+        if (didUserPassCB) {
           callback(err_obj);
+        } else {
+          deferred.reject(err_obj.error);
```
Comment (asisayag2) [inline lib-es5/api_client/execute_request.js]: We agreed to move all of those repeating 'if else' blocks with a 'handleError' method

File: lib/uploader.js [hunk 3]
```diff
@@ -355,11 +355,15 @@ function parseResult(buffer, res) {
   return result;
 }
 
-function call_api(action, callback, options, get_params) {
-  if (typeof callback !== "function") {
-    callback = function () {};
+function handleError(callback, result, deferred, error) {
```
Comment (patrick-tolosa) [inline lib/uploader.js]: There are a couple of these which are identical.

I think we should either rename to make them distinct or we can try to share them.
Which do you think makes more sense?

Comment (strausr) [reply lib/uploader.js]: updated

File: lib-es5/api_client/execute_request.js [hunk 4]
```diff
@@ -79,16 +87,17 @@ function execute_request(method, params, auth, api_url, callback) {
...
           }, result));
-        } else {
-          deferred.resolve(result);
         }
-        if (typeof callback === "function") {
+
+        if (callback) {
           callback(result);
+        } else {
+          deferred.resolve(result);
```
Comment (asisayag2) [inline lib-es5/api_client/execute_request.js]: Why not calling `handleRequestError`?

File: lib-es5/uploader.js [hunk 5]
```diff
@@ -463,14 +467,12 @@ function call_api(action, callback, options, get_params) {
...
   var handle_response = function handle_response(res) {
-    // let buffer;
     if (errorRaised) {
 
       // Already reported
     } else if (res.error) {
       errorRaised = true;
-      deferred.reject(res);
-      callback(res);
+      handleUploadAPIError(callback, res, deferred, res);
```
Comment (asisayag2) [inline lib-es5/uploader.js]: `res.error`?

File: lib-es5/uploader.js [hunk 6]
```diff
@@ -485,26 +487,27 @@ function call_api(action, callback, options, get_params) {
         result = parseResult(buffer, res);
         if (result.error) {
           result.error.http_code = res.statusCode;
-          deferred.reject(result.error);
+          handleUploadAPIError(callback, result, deferred, result.error);
```
Comment (asisayag2) [inline lib-es5/uploader.js]: @strausr , if the `error` variable is always `result.error`, why not sending only the result object?

File: lib-es5/uploader.js [hunk 7]
```diff
@@ -485,26 +487,27 @@ function call_api(action, callback, options, get_params) {
...
         } else {
           cacheResults(result, options);
-          deferred.resolve(result);
+          if (callback) {
+            callback(result);
+          } else {
+            deferred.resolve(result);
+          }
```
Comment (asisayag2) [inline lib-es5/uploader.js]: @strausr , why not calling the handleError?

File: lib-es5/uploader.js [hunk 8]
```diff
@@ -485,26 +487,27 @@ function call_api(action, callback, options, get_params) {
...
         }
-        callback(result);
       });
       res.on("error", function (error) {
         errorRaised = true;
-        deferred.reject(error);
-        callback({ error });
+        handleUploadAPIError(callback, { error }, deferred, error);
```
Comment (asisayag2) [inline lib-es5/uploader.js]: @strausr , why do we need the the {} wrapping?

File: lib-es5/api_client/execute_request.js (hunk 4 above)
Comment (strausr) [reply lib-es5/api_client/execute_request.js]: That's resolving.. it's not the same behavior

File: lib-es5/uploader.js (hunk 5 above)
Comment (strausr) [reply lib-es5/uploader.js]: No.. look at how the original error is sent

File: lib-es5/uploader.js (hunk 6 above)
Comment (strausr) [reply lib-es5/uploader.js]: in some places, we're just sending the result, like above.

File: lib-es5/uploader.js (hunk 7 above)
Comment (strausr) [reply lib-es5/uploader.js]: Resolving. We can make a resolving handler?

File: lib-es5/uploader.js (hunk 8 above)
Comment (strausr) [reply lib-es5/uploader.js]: Look at line 498 in the original

**Pull request closed without merging**